// @chakra-ui/c-skip-nav/dist/chakra-ui-c-skip-nav.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var vueSystem = require('@chakra-ui/vue-system');
var vueNext = require('@chakra-ui/vue-next');
var vueUtils = require('@chakra-ui/vue-utils');

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

var FALLBACK_ID = "chakra-skip-nav";
var CSkipNavLink = vue.defineComponent({
  name: "CSkipNavLink",
  props: {
    id: {
      type: String,
      "default": FALLBACK_ID
    }
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var styles = vueSystem.useStyleConfig("SkipLink", props);
    var skipLinkStyles = vue.computed(function () {
      return _extends({
        userSelect: "none",
        border: "0",
        borderRadius: "md",
        fontWeight: "semibold",
        height: "1px",
        width: "1px",
        margin: "-1px",
        padding: "0",
        outline: "0",
        overflow: "hidden",
        position: "absolute",
        clip: "rect(0 0 0 0)"
      }, styles.value, {
        _focus: {
          clip: "auto",
          width: "auto",
          height: "auto",
          boxShadow: "outline",
          padding: "1rem",
          position: "fixed",
          top: "1.5rem",
          insetStart: "1.5rem"
        }
      });
    });
    return function () {
      return vue.createVNode(vueSystem.chakra.a, vue.mergeProps({
        "href": "#" + props.id,
        "__css": skipLinkStyles.value
      }, attrs), {
        "default": function _default() {
          return vueUtils.getValidChildren(slots);
        }
      });
    };
  }
});
var CSkipNavContent = vue.defineComponent({
  name: "CSkipNavContent",
  props: {
    id: {
      type: String,
      "default": FALLBACK_ID
    }
  },
  setup: function setup(props, _ref2) {
    var attrs = _ref2.attrs,
        slots = _ref2.slots;
    return function () {
      return vue.createVNode(vueNext.CBox, vue.mergeProps({
        "tabIndex": "-1",
        "id": props.id,
        "data-testid": FALLBACK_ID
      }, attrs), {
        "default": function _default() {
          return vueUtils.getValidChildren(slots);
        }
      });
    };
  }
});

exports.CSkipNavContent = CSkipNavContent;
exports.CSkipNavLink = CSkipNavLink;
