// @chakra-ui/c-flex/dist/chakra-ui-c-flex.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var vueSystem = require('@chakra-ui/vue-system');

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

var CFlex = vue.defineComponent({
  props: {
    as: {
      type: [Object, String],
      "default": "div"
    },
    align: [Object, String, Array],
    justify: [Object, String, Array],
    wrap: [Object, String, Array],
    direction: {
      type: [Object, String, Array],
      "default": "row"
    },
    basis: [Object, String, Array],
    grow: [Object, String, Array],
    shrink: [Object, String, Array],
    size: String
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var styles = vue.reactive({
      display: "flex",
      flexDirection: props.direction,
      alignItems: props.align,
      justifyContent: props.justify,
      flexWrap: props.wrap,
      flexBasis: props.basis,
      flexGrow: props.grow,
      flexShrink: props.shrink,
      h: props.size,
      w: props.size
    });
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "as": props.as,
        "__label": "flex",
        "__css": styles
      }, attrs), _isSlot(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});
var CFlex$1 = CFlex;

exports.CFlex = CFlex$1;
