// @chakra-ui/vue-layout/dist/chakra-ui-vue-layout.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var utils = require('@chakra-ui/utils');
var vueSystem = require('@chakra-ui/vue-system');
var vueUtils = require('@chakra-ui/vue-utils');
var cIcon = require('@chakra-ui/c-icon');

function _isSlot$f(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

/**
 * Vue component used to cropping media (videos, images and maps)
 * to a desired aspect ratio.
 *
 * @see Docs https://vue.chakra-ui.com/docs/layout/aspect-ratio
 */
var CAspectRatio = vue.defineComponent({
  name: "CAspectRatio",
  props: {
    ratio: {
      type: [Number],
      "default": 4 / 3
    }
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "__label": "aspect-ratio",
        "position": "relative",
        "_before": {
          height: 0,
          content: "\"\"",
          display: "block",
          paddingBottom: utils.mapResponsive(props.ratio, function (r) {
            return 1 / r * 100 + "%";
          })
        },
        "__css": {
          "& > *:not(style)": {
            overflow: "hidden",
            position: "absolute",
            top: "0",
            right: "0",
            bottom: "0",
            left: "0",
            display: "flex",
            justifyContent: "center",
            alignItems: "center",
            width: "100%",
            height: "100%"
          },
          "& > img, & > video": {
            objectFit: "cover"
          }
        }
      }, attrs), _isSlot$f(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

function _isSlot$e(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

/**
 * Vue component used to display notifications, messages, or
 * statuses in different shapes and sizes.
 *
 * @see Docs https://vue.chakra-ui.com/docs/data-display/badge
 */
var CBadge = vue.defineComponent({
  name: "CBadge",
  props: _extends({
    as: {
      type: [Object, String],
      "default": "div"
    }
  }, vueUtils.vueThemingProps),
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var themingProps = vue.computed(function () {
      return utils.filterUndefined({
        colorScheme: props.colorScheme,
        variant: props.variant,
        size: props.size,
        styleConfig: props.styleConfig
      });
    });
    var styles = vueSystem.useStyleConfig("Badge", themingProps);
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "as": props.as,
        "__label": "badge",
        "__css": _extends({
          display: "inline-block",
          whiteSpace: "nowrap",
          verticalAlign: "middle"
        }, styles.value)
      }, attrs), _isSlot$e(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function _isSlot$d(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

/**
 * Box is the most abstract component on top of which other chakra
 * components are built. It renders a `div` element by default.
 *
 * @see Docs https://vue.chakra-ui.com/docs/layout/box
 */
var CBox = vue.defineComponent({
  name: "CBox",
  props: {
    as: {
      type: [String, Object],
      "default": "div"
    }
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "as": props.as,
        "__label": "box"
      }, attrs), {
        "default": function _default() {
          return slots == null ? void 0 : slots["default"] == null ? void 0 : slots["default"]();
        }
      });
    };
  }
});
/**
 * As a constraint, you can't pass size related props
 * Only `size` would be allowed
 */

/**
 * CSquare is the `CBox` component implemented as a square
 *
 * @see Docs https://vue.chakra-ui.com/docs/layout/box
 */
var CSquare = vue.defineComponent({
  name: "CSquare",
  props: {
    size: [Object, String, Number],
    centerContent: {
      type: [Boolean],
      "default": true
    }
  },
  setup: function setup(props, _ref2) {
    var slots = _ref2.slots,
        attrs = _ref2.attrs;
    var styles = vue.computed(function () {
      return props.centerContent ? {
        display: "flex",
        alignItems: "center",
        justifyContent: "center"
      } : {};
    });
    return function () {
      return vue.createVNode(CBox, vue.mergeProps({
        "__label": "square",
        "boxSize": props.size,
        "__css": _extends({}, styles.value, {
          flexShrink: 0,
          flexGrow: 0
        })
      }, attrs), _isSlot$d(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});
/**
 * CCircle is the `CBox` component implemented as a circle
 *
 * @see Docs https://vue.chakra-ui.com/docs/layout/box
 */

var CCircle = vue.defineComponent({
  name: "CCircle",
  setup: function setup(_, _ref3) {
    var slots = _ref3.slots,
        attrs = _ref3.attrs;
    return function () {
      return vue.createVNode(CSquare, vue.mergeProps({
        "__label": "circle",
        "borderRadius": "9999px"
      }, attrs), _isSlot$d(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function _isSlot$c(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

/**
 * Vue component used to horizontally and vertically center its child.
 * It uses the popular `display: flex` centering technique.
 *
 * @see Docs https://vue.chakra-ui.com/docs/layout/center
 */
var CCenter = vue.defineComponent({
  name: "CCenter",
  props: {
    as: {
      type: [String, Object],
      "default": "div"
    }
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "__label": "center",
        "__css": {
          display: "flex",
          alignItems: "center",
          justifyContent: "center"
        }
      }, props, attrs), _isSlot$c(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function _isSlot$b(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

/**
 * Layout component used to wrap app or website content
 *
 * It sets `margin-left` and `margin-right` to `auto`,
 * to keep its content centered.
 *
 * It also sets a default max-width of `60ch` (60 characters).
 */
var CContainer = vue.defineComponent({
  name: "CContainer",
  props: _extends({
    as: {
      type: [Object, String],
      "default": "div"
    },
    centerContent: {
      type: [Boolean]
    }
  }, vueUtils.vueThemingProps),
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var themingProps = vue.computed(function () {
      return utils.filterUndefined({
        colorScheme: props.colorScheme,
        variant: props.variant,
        size: props.size,
        styleConfig: props.styleConfig
      });
    });
    var styles = vueSystem.useStyleConfig("Container", themingProps);
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "__label": "container",
        "__css": _extends({}, styles.value, props.centerContent && {
          display: "flex",
          flexDirection: "column",
          alignItems: "center"
        })
      }, attrs), _isSlot$b(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

var _excluded = ["borderLeftWidth", "borderBottomWidth", "borderTopWidth", "borderRightWidth", "borderWidth", "borderStyle", "borderColor"];

/**
 * Layout component used to visually separate content in a list or group.
 * It display a thin horizontal or vertical line, and renders a `hr` tag.
 *
 * @see Docs https://vue.chakra-ui.com/docs/data-display/divider
 */
var CDivider = vue.defineComponent({
  name: "CDivider",
  props: _extends({
    orientation: {
      type: [String],
      "default": "horizontal"
    }
  }, vueUtils.vueThemingProps),
  setup: function setup(props, _ref) {
    var slots = _ref.slots;
        _ref.attrs;
    var themingProps = vue.computed(function () {
      return utils.filterUndefined({
        colorScheme: props.colorScheme,
        variant: props.variant,
        size: props.size,
        styleConfig: props.styleConfig,
        orientation: props.orientation
      });
    });
    var styles = vueSystem.useStyleConfig("Divider", themingProps);

    var _styles$value = styles.value,
        borderLeftWidth = _styles$value.borderLeftWidth,
        borderBottomWidth = _styles$value.borderBottomWidth,
        borderTopWidth = _styles$value.borderTopWidth,
        borderRightWidth = _styles$value.borderRightWidth,
        borderWidth = _styles$value.borderWidth,
        borderStyle = _styles$value.borderStyle,
        borderColor = _styles$value.borderColor,
        stylesRest = _objectWithoutPropertiesLoose(_styles$value, _excluded);

    var dividerStyle = vue.computed(function () {
      var dividerStyles = {
        vertical: {
          borderLeftWidth: borderLeftWidth || borderRightWidth || borderWidth || "1px",
          height: "100%"
        },
        horizontal: {
          borderBottomWidth: borderBottomWidth || borderTopWidth || borderWidth || "1px",
          width: "100%"
        }
      };
      return dividerStyles[props.orientation];
    });
    return function () {
      return vue.createVNode(vueSystem.chakra.hr, {
        "aria-orientation": props.orientation,
        "__css": _extends({}, stylesRest, {
          border: 0,
          borderColor: borderColor,
          borderStyle: borderStyle
        }, dividerStyle.value),
        "__label": "divider"
      }, {
        "default": function _default() {
          return [slots["default"] == null ? void 0 : slots["default"]()];
        }
      });
    };
  }
});

function _isSlot$a(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

/**
 * Vue component used to create grid layouts.
 *
 * It renders a `div` with `display: grid` and
 * comes with helpful style shorthand.
 *
 * @see Docs https://vue.chakra-ui.com/docs/layout/grid
 */
var CGrid = vue.defineComponent({
  name: "CGrid",
  props: {
    as: {
      type: [String, Object],
      "default": "div"
    },
    templateColumns: vueUtils.SNAO,
    gap: vueUtils.SNAO,
    rowGap: vueUtils.SNAO,
    columnGap: vueUtils.SNAO,
    autoFlow: vueUtils.SNAO,
    autoRows: vueUtils.SNAO,
    autoColumns: vueUtils.SNAO,
    templateRows: vueUtils.SNAO,
    templateAreas: vueUtils.SNAO,
    area: vueUtils.SNAO,
    column: vueUtils.SNAO,
    row: vueUtils.SNAO
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var styles = vue.computed(function () {
      return utils.filterUndefined({
        display: "grid",
        gridArea: props.area,
        gridTemplateAreas: props.templateAreas,
        gridGap: props.gap,
        gridRowGap: props.rowGap,
        gridColumnGap: props.columnGap,
        gridAutoColumns: props.autoColumns,
        gridColumn: props.column,
        gridRow: props.row,
        gridAutoFlow: props.autoFlow,
        gridAutoRows: props.autoRows,
        gridTemplateRows: props.templateRows,
        gridTemplateColumns: props.templateColumns
      });
    });
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "as": props.as,
        "__label": "grid",
        "__css": styles.value
      }, attrs), _isSlot$a(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function spanFn(span) {
  return utils.mapResponsive(span, function (value) {
    return value === "auto" ? "auto" : "span " + value + "/span " + value;
  });
}

var CGridItem = vue.defineComponent({
  name: "CGridItem",
  props: {
    as: {
      type: [String, Object],
      "default": "div"
    },
    colSpan: vueUtils.SNAO,
    colStart: vueUtils.SNAO,
    colEnd: vueUtils.SNAO,
    rowStart: vueUtils.SNAO,
    rowEnd: vueUtils.SNAO,
    rowSpan: vueUtils.SNAO
  },
  setup: function setup(props, _ref2) {
    var slots = _ref2.slots,
        attrs = _ref2.attrs;
    var styles = vue.computed(function () {
      return utils.filterUndefined({
        gridColumn: spanFn(props.colSpan),
        gridRow: spanFn(props.rowSpan),
        gridColumnStart: props.colStart,
        gridColumnEnd: props.colEnd,
        gridRowStart: props.rowStart,
        gridRowEnd: props.rowEnd
      });
    });
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "as": props.as,
        "__label": "grid__item",
        "__css": styles.value
      }, attrs), _isSlot$a(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function _isSlot$9(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

var CHeading = vue.defineComponent({
  name: "CHeading",
  props: _extends({
    as: {
      type: [String, Object],
      "default": "h2"
    }
  }, vueUtils.vueThemingProps),
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var themingProps = vue.computed(function () {
      return utils.filterUndefined({
        colorScheme: props.colorScheme,
        variant: props.variant,
        size: props.size,
        styleConfig: props.styleConfig
      });
    });
    var styles = vueSystem.useStyleConfig("Heading", themingProps);
    return function () {
      return vue.createVNode(vueSystem.chakra.h2, vue.mergeProps({
        "as": props.as,
        "__label": "heading",
        "__css": styles.value
      }, attrs), _isSlot$9(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function _isSlot$8(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

/**
 * Links are accessible elements used primarily for navigation.
 *
 * It integrates well with other routing libraries like
 * Vue Router and Nuxt.js Link.
 *
 * @example
 *
 * ```vue
 * <CLink as="router-link" to="/home">Home</CLink>
 * ```
 *
 * @see Docs https://vue.chakra-ui.com/docs/layout/link
 */
var CLink = vue.defineComponent({
  name: "CLink",
  props: _extends({
    as: {
      type: [Object, String],
      "default": "a"
    },
    isExternal: Boolean
  }, vueUtils.vueThemingProps),
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var themingProps = vue.computed(function () {
      return utils.filterUndefined({
        colorScheme: props.colorScheme,
        variant: props.variant,
        size: props.size,
        styleConfig: props.styleConfig
      });
    });
    var styles = vueSystem.useStyleConfig("Link", themingProps);
    return function () {
      return vue.createVNode(vueSystem.chakra.a, vue.mergeProps({
        "as": props.as,
        "__label": "link",
        "target": props.isExternal ? "_blank" : undefined,
        "rel": props.isExternal ? "noopener noreferrer" : undefined,
        "__css": styles.value
      }, attrs), _isSlot$8(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function _isSlot$7(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

var CLinkOverlay = vue.defineComponent({
  name: "CLinkOverlay",
  props: {
    as: {
      type: [Object, String],
      "default": "a"
    },
    isExternal: Boolean
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "as": props.as,
        "__label": "linkbox__overlay",
        "rel": props.isExternal ? "noopener noreferrer" : undefined,
        "target": props.isExternal ? "_blank" : undefined,
        "__css": {
          position: "static",
          "&::before": {
            content: "''",
            cursor: "inherit",
            display: "block",
            position: "absolute",
            top: 0,
            left: 0,
            zIndex: 0,
            width: "100%",
            height: "100%"
          }
        }
      }, attrs), _isSlot$7(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

/**
 * `LinkBox` is used to wrap content areas within a link while ensuring semantic html
 *
 * @see Docs https://vue.chakra-ui.com/docs/link-overlay
 * @see Resources https://www.sarasoueidan.com/blog/nested-links
 */
var CLinkBox = vue.defineComponent({
  name: "CLinkBox",
  props: {
    as: {
      type: [Object, String],
      "default": "div"
    }
  },
  setup: function setup(props, _ref2) {
    var slots = _ref2.slots,
        attrs = _ref2.attrs;
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "as": props.as,
        "__label": "linkbox",
        "position": "relative",
        "__css": {
          /* Elevates links and abbreviations */
          "a[href]:not(.chakra-linkbox__overlay), abbr[title]": {
            position: "relative",
            zIndex: 1
          }
        }
      }, attrs), _isSlot$7(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function _isSlot$6(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

/**
 * List is used to display list items, it renders a `<ul>` by default.
 *
 * @see Docs https://vue.chakra-ui.com/docs/data-display/list
 */
var CList = vue.defineComponent({
  name: "CList",
  props: {
    as: {
      type: [Object, String],
      "default": "ul"
    },
    styleType: {
      type: vueUtils.SAO,
      "default": "none"
    },
    stylePosition: vueUtils.SAO,
    spacing: vueUtils.SNAO
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var styles = vueSystem.useMultiStyleConfig("List", props);
    vueSystem.StylesProvider(styles);
    var selector = "& > *:not(style) ~ *:not(style)";
    var spacingStyle = vue.computed(function () {
      var _ref2;

      return props.spacing ? (_ref2 = {}, _ref2[selector] = {
        mt: props.spacing
      }, _ref2) : {};
    });
    return function () {
      var validChildren = function validChildren() {
        return vueUtils.getValidChildren(slots);
      };

      return vue.createVNode(vueSystem.chakra.ul, vue.mergeProps({
        "__label": "list",
        "as": props.as,
        "listStyleType": props.styleType,
        "listStylePosition": props.stylePosition,
        "role": "list",
        "__css": _extends({}, styles.value.container, spacingStyle.value)
      }, attrs), _isSlot$6(validChildren) ? validChildren : {
        "default": function _default() {
          return [validChildren];
        }
      });
    };
  }
});
var COrderedList = vue.defineComponent({
  name: "COrderedList",
  setup: function setup(props, _ref3) {
    var slots = _ref3.slots,
        attrs = _ref3.attrs;
    return function () {
      return (// @ts-ignore
        vue.createVNode(CList, vue.mergeProps({
          "styleType": "decimal",
          "marginStart": "1em"
        }, attrs), _isSlot$6(slots) ? slots : {
          "default": function _default() {
            return [slots];
          }
        })
      );
    };
  }
});
var CUnorderedList = vue.defineComponent({
  name: "CUnorderedList",
  setup: function setup(props, _ref4) {
    var slots = _ref4.slots,
        attrs = _ref4.attrs;
    return function () {
      return (// @ts-ignore
        vue.createVNode(CList, vue.mergeProps({
          "styleType": "initial",
          "marginStart": "1em"
        }, attrs), _isSlot$6(slots) ? slots : {
          "default": function _default() {
            return [slots];
          }
        })
      );
    };
  }
});
var CListItem = vue.defineComponent({
  name: "CListItem",
  setup: function setup(_, _ref5) {
    var slots = _ref5.slots,
        attrs = _ref5.attrs;
    var styles = vueSystem.useStyles();
    return function () {
      return vue.createVNode(vueSystem.chakra.li, vue.mergeProps({
        "__label": "list__item",
        "__css": styles.value.item
      }, attrs), _isSlot$6(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});
var CListIcon = vue.defineComponent({
  name: "CListIcon",
  setup: function setup(_, _ref6) {
    var slots = _ref6.slots,
        attrs = _ref6.attrs;
    var styles = vueSystem.useStyles();
    return function () {
      return vue.createVNode(cIcon.CIcon, vue.mergeProps({
        "role": "presentation"
      }, attrs, {
        "__css": styles.value.icon
      }), _isSlot$6(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function _isSlot$5(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

/**
 * Semantic component to render a keyboard shortcut
 * within an application.
 *
 * @example
 *
 * ```jsx
 * <CKbd>⌘ + T</CKbd>
 * ```
 *
 * @see Docs https://vue.chakra-ui.com/docs/data-display/kbd
 */
var CKbd = vue.defineComponent({
  name: "CKbd",
  props: _extends({
    as: {
      type: [String, Object],
      "default": "h2"
    }
  }, vueUtils.vueThemingProps),
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var themingProps = vue.computed(function () {
      return utils.filterUndefined({
        colorScheme: props.colorScheme,
        variant: props.variant,
        size: props.size,
        styleConfig: props.styleConfig
      });
    });
    var styles = vueSystem.useStyleConfig("Kbd", themingProps);
    return function () {
      return vue.createVNode(vueSystem.chakra.kbd, vue.mergeProps({
        "__label": "kbd",
        "__css": _extends({
          fontFamily: "mono"
        }, styles.value)
      }, attrs), _isSlot$5(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function _isSlot$4(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

/**
 * SimpleGrid
 *
 * Vue component make that providers a simpler interface, and
 * make its easy to create responsive grid layouts.
 *
 * @see Docs https://vue.chakra-ui.com/docs/layout/simple-grid
 */
var CSimpleGrid = vue.defineComponent({
  name: "CSimpleGrid",
  props: {
    as: {
      type: [Object, String],
      "default": "ul"
    },
    minChildWidth: vueUtils.SNAO,
    columns: vueUtils.SNAO,
    spacing: vueUtils.SNAO,
    spacingX: vueUtils.SNAO,
    spacingY: vueUtils.SNAO
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var templateColumns = vue.computed(function () {
      return props.minChildWidth ? widthToColumns(props.minChildWidth) : countToColumns(props.columns);
    });
    return function () {
      return vue.createVNode(CGrid, vue.mergeProps({
        "as": props.as,
        "__label": "simple-grid",
        "gap": props.spacing,
        "columnGap": props.spacingX,
        "rowGap": props.spacingY,
        "templateColumns": templateColumns.value
      }, attrs), _isSlot$4(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function toPx(n) {
  return utils.isNumber(n) ? n + "px" : n;
}

function widthToColumns(width) {
  return utils.mapResponsive(width, function (value) {
    return utils.isNull(value) ? null : "repeat(auto-fit, minmax(" + toPx(value) + ", 1fr))";
  });
}

function countToColumns(count) {
  return utils.mapResponsive(count, function (value) {
    return utils.isNull(value) ? null : "repeat(" + value + ", minmax(0, 1fr))";
  });
}

function _isSlot$3(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

/**
 * A flexible flex spacer that expands along the major axis of its containing flex layout.
 * It renders a `div` by default, and takes up any available space.
 *
 * @see Docs https://chakra-ui.com/docs/layout/flex#using-the-spacer
 */
var CSpacer = vue.defineComponent({
  name: "CSpacer",
  setup: function setup(_, _ref) {
    var slots = _ref.slots;
    return function () {
      return vue.createVNode(vueSystem.chakra.div, {
        "__label": "spacer",
        "baseStyle": {
          flex: 1,
          justifySelf: "stretch",
          alignSelf: "stretch"
        }
      }, _isSlot$3(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

/**
 * If we ever run into SSR issues with this, check this post to find a fix for it:
 * @see https://medium.com/@emmenko/patching-lobotomized-owl-selector-for-emotion-ssr-5a582a3c424c
 */
var selector = "& > *:not(style) ~ *:not(style)";
function getStackStyles(options) {
  var _ref;

  var spacing = options.spacing,
      direction = options.direction;
  var directionStyles = {
    column: {
      marginTop: spacing,
      marginEnd: 0,
      marginBottom: 0,
      marginStart: 0
    },
    row: {
      marginTop: 0,
      marginEnd: 0,
      marginBottom: 0,
      marginStart: spacing
    },
    "column-reverse": {
      marginTop: 0,
      marginEnd: 0,
      marginBottom: spacing,
      marginStart: 0
    },
    "row-reverse": {
      marginTop: 0,
      marginEnd: spacing,
      marginBottom: 0,
      marginStart: 0
    }
  };
  return _ref = {
    flexDirection: direction
  }, _ref[selector] = utils.mapResponsive(direction, function (value) {
    return directionStyles[value];
  }), _ref;
}
function getDividerStyles(options) {
  var spacing = options.spacing,
      direction = options.direction;
  var dividerStyles = {
    column: {
      my: spacing,
      mx: 0,
      borderLeftWidth: 0,
      borderBottomWidth: "1px"
    },
    "column-reverse": {
      my: spacing,
      mx: 0,
      borderLeftWidth: 0,
      borderBottomWidth: "1px"
    },
    row: {
      mx: spacing,
      my: 0,
      borderLeftWidth: "1px",
      borderBottomWidth: 0
    },
    "row-reverse": {
      mx: spacing,
      my: 0,
      borderLeftWidth: "1px",
      borderBottomWidth: 0
    }
  };
  return {
    "&": utils.mapResponsive(direction, function (value) {
      return dividerStyles[value];
    })
  };
}

function _isSlot$2(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

var CStackDivider = vue.defineComponent({
  name: "CStackDivider",
  inheritAttrs: false,
  setup: function setup(_, _ref) {
    var attrs = _ref.attrs,
        slots = _ref.slots;
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "__label": "stack__divider",
        "borderWidth": 0,
        "alignSelf": "stretch",
        "borderColor": "inherit",
        "width": "auto",
        "height": "auto"
      }, attrs), {
        "default": function _default() {
          return [slots == null ? void 0 : slots["default"] == null ? void 0 : slots["default"]()];
        }
      });
    };
  }
});
var CStackItem = vue.defineComponent({
  name: "CStackItem",
  setup: function setup(_, _ref2) {
    var attrs = _ref2.attrs,
        slots = _ref2.slots;
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "__label": "stack__item",
        "display": "inline-block",
        "flex": "0 0 auto",
        "minWidth": "0"
      }, attrs), {
        "default": function _default() {
          return [slots == null ? void 0 : slots["default"] == null ? void 0 : slots["default"]()];
        }
      });
    };
  }
});
var stackProps = {
  as: {
    type: [Object, String],
    "default": "div"
  },
  align: vueUtils.SAO,
  justify: vueUtils.SAO,
  wrap: vueUtils.SAO,
  spacing: {
    type: vueUtils.SNAO,
    "default": "0.5rem"
  },
  direction: vueUtils.SAO,
  // todo: divider
  divider: [Object, Boolean],
  shouldWrapChildren: [Boolean],
  isInline: [Boolean]
};
/**
 * Stacks help you easily create flexible and automatically distributed layouts
 *
 * You can stack elements in the horizontal or vertical direction,
 * and apply a space or/and divider between each element.
 *
 * It uses `display: flex` internally and renders a `div`.
 *
 * @see Docs https://vue.chakra-ui.com/docs/layout/stack
 *
 */

var CStack = vue.defineComponent({
  name: "CStack",
  props: stackProps,
  setup: function setup(props, _ref3) {
    var slots = _ref3.slots,
        attrs = _ref3.attrs;
    var direction = vue.computed(function () {
      var _props$direction;

      return props.isInline ? "row" : (_props$direction = props.direction) != null ? _props$direction : "column";
    });
    var styles = vue.computed(function () {
      return getStackStyles({
        direction: direction.value,
        spacing: props.spacing
      });
    });
    var dividerStyle = vue.computed(function () {
      return getDividerStyles({
        spacing: props.spacing,
        direction: direction.value
      });
    });
    var hasDivider = vue.computed(function () {
      return !!props.divider;
    });
    var shouldUseChildren = vue.computed(function () {
      return !props.shouldWrapChildren && !hasDivider.value;
    });
    return function () {
      var _ref4;

      var validChildren = vueUtils.getValidChildren(slots);
      var clones = shouldUseChildren.value ? validChildren : validChildren.map(function (child, index) {
        var isLast = index + 1 === validChildren.length;
        var wrappedChild = vue.createVNode(CStackItem, {
          key: index
        }, child);

        var _child = props.shouldWrapChildren ? wrappedChild : child;

        if (!hasDivider.value) return _child; // todo: temporary divider

        var clonedDivider = vue.createVNode(CStackDivider, {
          borderColor: "blue.200",
          __css: dividerStyle.value
        });

        var _divider = isLast ? null : clonedDivider;

        return vue.createVNode(vue.Fragment, {
          key: index
        }, [_child, _divider]);
      });
      return vue.createVNode(vueSystem.chakra.div, {
        "__label": attrs.label ? attrs.label : "stack",
        "display": "flex",
        "alignItems": props.align,
        "justifyContent": props.justify,
        "flexDirection": styles.value.flexDirection,
        "flexWrap": props.wrap,
        "__css": hasDivider.value ? {} : (_ref4 = {}, _ref4[selector] = styles.value[selector], _ref4)
      }, {
        "default": function _default() {
          return clones;
        }
      });
    };
  }
});
/**
 * A view that arranges its children in a horizontal line.
 */

var CHStack = vue.defineComponent({
  name: "CHStack",
  props: stackProps,
  setup: function setup(props, _ref5) {
    var attrs = _ref5.attrs,
        slots = _ref5.slots;
    return function () {
      return (// @ts-expect-error Stack typed API
        vue.createVNode(CStack, vue.mergeProps({
          "__label": "stack-horizontal"
        }, props, attrs, {
          "direction": "row"
        }), _isSlot$2(slots) ? slots : {
          "default": function _default() {
            return [slots];
          }
        })
      );
    };
  }
});
/**
 * A view that arranges its children in a vertical line.
 */

var CVStack = vue.defineComponent({
  name: "CVStack",
  props: stackProps,
  setup: function setup(props, _ref6) {
    var attrs = _ref6.attrs,
        slots = _ref6.slots;
    return function () {
      return (// @ts-expect-error Stack typed API
        vue.createVNode(CStack, vue.mergeProps({
          "__label": "stack-vertical"
        }, props, attrs, {
          "direction": "column"
        }), _isSlot$2(slots) ? slots : {
          "default": function _default() {
            return [slots];
          }
        })
      );
    };
  }
});

function _isSlot$1(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

/**
 * Used to render texts or paragraphs.
 *
 * @see Docs https://vue.chakra-ui.com/docs/typography/text
 */
var CText = vue.defineComponent({
  name: "CText",
  props: _extends({
    as: {
      type: [Object, String],
      "default": "p"
    },
    align: vueUtils.SNAO,
    decoration: vueUtils.SNAO,
    casing: vueUtils.SNAO
  }, vueUtils.vueThemingProps),
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var themingProps = vue.computed(function () {
      return utils.filterUndefined({
        colorScheme: props.colorScheme,
        variant: props.variant,
        size: props.size,
        styleConfig: props.styleConfig
      });
    });
    var styles = vueSystem.useStyleConfig("Text", themingProps);
    var aliasedProps = vue.computed(function () {
      return utils.filterUndefined({
        textAlign: props.align,
        textDecoration: props.decoration,
        textTransform: props.casing
      });
    });
    return function () {
      return vue.createVNode(vueSystem.chakra.p, vue.mergeProps({
        "__label": "text"
      }, aliasedProps.value, {
        "__css": styles.value
      }, attrs), _isSlot$1(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

var CWrapProps = {
  spacing: vueUtils.SNAO,
  justify: vueUtils.SNAO,
  align: vueUtils.SNAO,
  direction: vueUtils.SNAO,
  shouldWrapChildren: vueUtils.SNAO
};
/**
 * Used to render texts or paragraphs.
 *
 * @see Docs https://vue.chakra-ui.com/docs/typography/text
 */

var CWrap = vue.defineComponent({
  props: _extends({
    as: {
      type: [Object, String],
      "default": "div"
    }
  }, CWrapProps),
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var styles = vue.computed(function () {
      return {
        "--chakra-wrap-spacing": function chakraWrapSpacing(theme) {
          return utils.mapResponsive(props.spacing, function (value) {
            return vueSystem.tokenToCSSVar("space", value)(theme);
          });
        },
        "--wrap-spacing": "calc(var(--chakra-wrap-spacing) / 2)",
        display: "flex",
        flexWrap: "wrap",
        justifyContent: props.justify,
        alignItems: props.align,
        flexDirection: props.direction,
        listStyleType: "none",
        padding: "0",
        margin: "calc(var(--wrap-spacing) * -1)",
        "& > *:not(style)": {
          margin: "var(--wrap-spacing)"
        }
      };
    });
    var childrenToRender = props.shouldWrapChildren ? vueUtils.getValidChildren(slots).map(function (child, index) {
      return vue.createVNode(CWrapItem, {
        "key": index
      }, _isSlot(child) ? child : {
        "default": function _default() {
          return [child];
        }
      });
    }) : slots;
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "as": props.as,
        "__label": "wrap"
      }, attrs), {
        "default": function _default() {
          return vue.createVNode(vueSystem.chakra.ul, {
            "__label": "wrap__list",
            "__css": styles.value
          }, _isSlot(childrenToRender) ? childrenToRender : {
            "default": function _default() {
              return [childrenToRender];
            }
          });
        }
      });
    };
  }
});
var CWrapItem = vue.defineComponent({
  setup: function setup(_, _ref2) {
    var attrs = _ref2.attrs,
        slots = _ref2.slots;
    return function () {
      return vue.createVNode(vueSystem.chakra.li, vue.mergeProps({
        "__label": "wrap__listItem",
        "__css": {
          display: "flex",
          alignItems: "flex-start"
        }
      }, attrs), {
        "default": function _default() {
          return vueUtils.getValidChildren(slots);
        }
      });
    };
  }
});

exports.CAspectRatio = CAspectRatio;
exports.CBadge = CBadge;
exports.CBox = CBox;
exports.CCenter = CCenter;
exports.CCircle = CCircle;
exports.CContainer = CContainer;
exports.CDivider = CDivider;
exports.CGrid = CGrid;
exports.CGridItem = CGridItem;
exports.CHStack = CHStack;
exports.CHeading = CHeading;
exports.CKbd = CKbd;
exports.CLink = CLink;
exports.CLinkBox = CLinkBox;
exports.CLinkOverlay = CLinkOverlay;
exports.CList = CList;
exports.CListIcon = CListIcon;
exports.CListItem = CListItem;
exports.COrderedList = COrderedList;
exports.CSimpleGrid = CSimpleGrid;
exports.CSpacer = CSpacer;
exports.CSquare = CSquare;
exports.CStack = CStack;
exports.CStackDivider = CStackDivider;
exports.CStackItem = CStackItem;
exports.CText = CText;
exports.CUnorderedList = CUnorderedList;
exports.CVStack = CVStack;
exports.CWrap = CWrap;
exports.CWrapItem = CWrapItem;
exports.CWrapProps = CWrapProps;
