// @chakra-ui/c-tag/dist/chakra-ui-c-tag.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var vueSystem = require('@chakra-ui/vue-system');
var cIcon = require('@chakra-ui/c-icon');
var utils = require('@chakra-ui/utils');
var vueUtils = require('@chakra-ui/vue-utils');

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

var CTagLabel = vue.defineComponent({
  props: _extends({}, vueUtils.vueThemingProps),
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var themingProps = vue.computed(function () {
      return utils.filterUndefined({
        colorScheme: props.colorScheme,
        variant: props.variant,
        size: props.size,
        styleConfig: props.styleConfig
      });
    });
    var styles = vueSystem.useMultiStyleConfig("Tag", themingProps);
    return function () {
      return vue.createVNode(vueSystem.chakra.span, vue.mergeProps({
        "__css": styles.value.label,
        "noOfLines": 1
      }, attrs), {
        "default": function _default() {
          return vueUtils.getValidChildren(slots);
        }
      });
    };
  }
});

var TagProps = _extends({
  variantColor: String
}, vueUtils.vueThemingProps);

var CTagLeftIcon = vue.defineComponent({
  setup: function setup(props, _ref2) {
    var attrs = _ref2.attrs;
    return function () {
      return vue.createVNode(cIcon.CIcon, vue.mergeProps(attrs, {
        "marginEnd": "0.5rem"
      }), null);
    };
  }
});
var CTagRightIcon = vue.defineComponent({
  setup: function setup(props, _ref3) {
    var attrs = _ref3.attrs;
    return function () {
      return vue.createVNode(cIcon.CIcon, vue.mergeProps(attrs, {
        "marginStart": "0.5rem"
      }), null);
    };
  }
});

var CloseButtonProps = _extends({
  isDisabled: Boolean
}, vueUtils.vueThemingProps);

var CTagCloseButton = vue.defineComponent({
  props: CloseButtonProps,
  setup: function setup(props, _ref4) {
    var slots = _ref4.slots,
        attrs = _ref4.attrs;
    var themingProps = vue.computed(function () {
      return utils.filterUndefined({
        colorScheme: props.colorScheme,
        variant: props.variant,
        size: props.size,
        styleConfig: props.styleConfig
      });
    });
    var styles = vueSystem.useMultiStyleConfig("Tag", themingProps);

    var buttonStyles = _extends({
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      outline: "0"
    }, styles.value.closeButton);

    return function () {
      return vue.createVNode(vueSystem.chakra.button, vue.mergeProps({
        "aria-label": "close",
        "type": "button",
        "__css": buttonStyles
      }, attrs, {
        "disabled": props == null ? void 0 : props.isDisabled
      }), {
        "default": function _default() {
          return [slots["default"] ? function () {
            return vueUtils.getValidChildren(slots);
          } : function () {
            return vue.createVNode(cIcon.CIcon, {
              "name": "close"
            }, null);
          }];
        }
      });
    };
  }
});
var CTag = vue.defineComponent({
  props: TagProps,
  setup: function setup(props, _ref5) {
    var slots = _ref5.slots,
        attrs = _ref5.attrs;
    var themingProps = vue.computed(function () {
      return utils.filterUndefined({
        colorScheme: props.colorScheme,
        variant: props.variant,
        size: props.size,
        styleConfig: props.styleConfig
      });
    });
    var styles = vueSystem.useMultiStyleConfig("Tag", themingProps);
    var tagContainerStyles = vue.computed(function () {
      var _styles$value, _props$variantColor, _styles$value2, _styles$value2$contai;

      return _extends({}, (_styles$value = styles.value) == null ? void 0 : _styles$value.container, {
        bg: (_props$variantColor = props.variantColor) != null ? _props$variantColor : (_styles$value2 = styles.value) == null ? void 0 : (_styles$value2$contai = _styles$value2.container) == null ? void 0 : _styles$value2$contai.bg
      });
    });
    return function () {
      return vue.createVNode(vueSystem.chakra.span, vue.mergeProps({
        "__label": "tag",
        "aria-label": "tag",
        "__css": tagContainerStyles.value
      }, attrs), {
        "default": function _default() {
          return vueUtils.getValidChildren(slots);
        }
      });
    };
  }
});

exports.CTag = CTag;
exports.CTagCloseButton = CTagCloseButton;
exports.CTagLabel = CTagLabel;
exports.CTagLeftIcon = CTagLeftIcon;
exports.CTagRightIcon = CTagRightIcon;
