// @chakra-ui/vue-anatomy/dist/chakra-ui-vue-anatomy.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vueThemeTools = require('@chakra-ui/vue-theme-tools');

/**
 * Hey! Welcome to @chakra-ui/vue-next Anatomy
 *
 * The anatomy of all chakra components
 *
 * @see Docs     https://next.vue.chakra-ui.com/vue-anatomy
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue-next/blob/master/packages/vue-anatomy/src/vue-anatomy/vue-anatomy.ts
 * @see WAI-ARIA https://www.w3.org/TR/wai-aria-practices-1.2
 */
/**
 * **Accordion anatomy**
 * - Root: the root container of the accordion
 * - Container: the accordion item contains the button and panel
 * - Button: the button is the trigger for the panel
 * - Panel: the panel is the content of the accordion item
 * - Icon: the expanded/collapsed icon
 */

var accordionAnatomy = vueThemeTools.anatomy("accordion").parts("root", "container", "button", "panel").extend("icon");
/**
 * **Alert anatomy**
 * - Title: the alert's title
 * - Description: the alert's description
 * - Icon: the alert's icon
 */

var alertAnatomy = vueThemeTools.anatomy("alert").parts("title", "description", "container").extend("icon", "spinner");
/**
 * **Avatar anatomy**
 * - Container: the container for the avatar
 * - Label: the avatar initials text
 * - Excess Label: the label or text that represents excess avatar count.
 * Typically used in avatar groups.
 * - Group: the container for the avatar group
 */

var avatarAnatomy = vueThemeTools.anatomy("avatar").parts("label", "badge", "container").extend("excessLabel", "group");
/**
 * **Breadcrumb anatomy**
 * - Item: the container for a breadcrumb item
 * - Link: the element that represents the breadcrumb link
 * - Container: the container for the breadcrumb items
 * - Separator: the separator between breadcrumb items
 */

var breadcrumbAnatomy = vueThemeTools.anatomy("breadcrumb").parts("link", "item", "container").extend("separator");
var buttonAnatomy = vueThemeTools.anatomy("button").parts();
var checkboxAnatomy = vueThemeTools.anatomy("checkbox").parts("control", "icon", "container").extend("label");
var circularProgressAnatomy = vueThemeTools.anatomy("progress").parts("track", "filledTrack").extend("label");
var drawerAnatomy = vueThemeTools.anatomy("drawer").parts("overlay", "dialogContainer", "dialog").extend("header", "closeButton", "body", "footer");
var editableAnatomy = vueThemeTools.anatomy("editable").parts("preview", "input", "textarea");
var formAnatomy = vueThemeTools.anatomy("form").parts("container", "requiredIndicator", "helperText");
var formErrorAnatomy = vueThemeTools.anatomy("formError").parts("text", "icon");
var inputAnatomy = vueThemeTools.anatomy("input").parts("addon", "field", "element");
var listAnatomy = vueThemeTools.anatomy("list").parts("container", "item", "icon");
var menuAnatomy = vueThemeTools.anatomy("menu").parts("button", "list", "item").extend("groupTitle", "command", "divider");
var modalAnatomy = vueThemeTools.anatomy("modal").parts("overlay", "dialogContainer", "dialog").extend("header", "closeButton", "body", "footer");
var numberInputAnatomy = vueThemeTools.anatomy("numberinput").parts("root", "field", "stepperGroup", "stepper");
var pinInputAnatomy = vueThemeTools.anatomy("pininput").parts("field");
var popoverAnatomy = vueThemeTools.anatomy("popover").parts("content", "header", "body", "footer").extend("popper", "arrow", "closeButton");
var progressAnatomy = vueThemeTools.anatomy("progress").parts("label", "filledTrack", "track");
var radioAnatomy = vueThemeTools.anatomy("radio").parts("container", "control", "label");
var selectAnatomy = vueThemeTools.anatomy("select").parts("field", "icon");
var sliderAnatomy = vueThemeTools.anatomy("slider").parts("container", "track", "thumb", "filledTrack");
var statAnatomy = vueThemeTools.anatomy("stat").parts("container", "label", "helpText", "number", "icon");
var switchAnatomy = vueThemeTools.anatomy("switch").parts("container", "track", "thumb");
var tableAnatomy = vueThemeTools.anatomy("table").parts("table", "thead", "tbody", "tr", "th", "td", "tfoot", "caption");
var tabsAnatomy = vueThemeTools.anatomy("tabs").parts("root", "tab", "tablist", "tabpanel", "tabpanels", "indicator");
/**
 * **Tag anatomy**
 * - Container: the container for the tag
 * - Label: the text content of the tag
 * - closeButton: the close button for the tag
 */

var tagAnatomy = vueThemeTools.anatomy("tag").parts("container", "label", "closeButton");

exports.accordionAnatomy = accordionAnatomy;
exports.alertAnatomy = alertAnatomy;
exports.avatarAnatomy = avatarAnatomy;
exports.breadcrumbAnatomy = breadcrumbAnatomy;
exports.buttonAnatomy = buttonAnatomy;
exports.checkboxAnatomy = checkboxAnatomy;
exports.circularProgressAnatomy = circularProgressAnatomy;
exports.drawerAnatomy = drawerAnatomy;
exports.editableAnatomy = editableAnatomy;
exports.formAnatomy = formAnatomy;
exports.formErrorAnatomy = formErrorAnatomy;
exports.inputAnatomy = inputAnatomy;
exports.listAnatomy = listAnatomy;
exports.menuAnatomy = menuAnatomy;
exports.modalAnatomy = modalAnatomy;
exports.numberInputAnatomy = numberInputAnatomy;
exports.pinInputAnatomy = pinInputAnatomy;
exports.popoverAnatomy = popoverAnatomy;
exports.progressAnatomy = progressAnatomy;
exports.radioAnatomy = radioAnatomy;
exports.selectAnatomy = selectAnatomy;
exports.sliderAnatomy = sliderAnatomy;
exports.statAnatomy = statAnatomy;
exports.switchAnatomy = switchAnatomy;
exports.tableAnatomy = tableAnatomy;
exports.tabsAnatomy = tabsAnatomy;
exports.tagAnatomy = tagAnatomy;
