// @chakra-ui/c-spinner/dist/chakra-ui-c-spinner.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var vueUtils = require('@chakra-ui/vue-utils');
var vueSystem = require('@chakra-ui/vue-system');
var cVisuallyHidden = require('@chakra-ui/c-visually-hidden');
var utils = require('@chakra-ui/utils');

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

var spin = vueSystem.keyframes({
  "0%": {
    transform: "rotate(0deg)"
  },
  "100%": {
    transform: "rotate(360deg)"
  }
});
var defaultSpinnerProps = {
  as: "div",
  emptyColor: "transparent",
  thickness: "2px",
  speed: "0.45s"
};
var CSpinner = vue.defineComponent({
  props: _extends({
    as: vueUtils.SNAO,

    /**
     * The color of the empty area in the spinner
     */
    emptyColor: vueUtils.SNAO,

    /**
     * The color of the spinner
     */
    color: vueUtils.SNAO,

    /**
     * The thickness of the spinner
     * @example
     * ```html
     * <c-spinner thickness="4px"/>
     * ```
     */
    thickness: vueUtils.SNAO,

    /**
     * The speed of the spinner.
     * @example
     * ```html
     * <c-spinner speed="0.2s"/>
     * ```
     */
    speed: vueUtils.SNAO,

    /**
     * For accessibility, it is important to add a fallback loading text.
     * This text will be visible to screen readers.
     */
    label: vueUtils.SNAO
  }, vueUtils.vueThemingProps),
  setup: function setup(_props, _ref) {
    _ref.slots;
        var attrs = _ref.attrs;
    var props = vue.computed(function () {
      return utils.mergeWith({}, defaultSpinnerProps, _props);
    });
    var themingProps = vue.computed(function () {
      return {
        colorScheme: props.value.colorScheme,
        variant: props.value.variant,
        size: props.value.size,
        styleConfig: props.value.styleConfig
      };
    });
    var styles = vueSystem.useStyleConfig("Spinner", themingProps);
    var spinnerStyles = vue.computed(function () {
      return _extends({
        display: "inline-block",
        borderColor: "currentColor",
        borderStyle: "solid",
        borderRadius: "99999px",
        borderWidth: props.value.thickness,
        borderBottomColor: props.value.emptyColor,
        borderLeftColor: props.value.emptyColor,
        color: props.value.color,
        animation: spin + " " + props.value.speed + " linear infinite"
      }, styles.value);
    });
    return function () {
      return vue.createVNode(vueSystem.chakra.div, vue.mergeProps({
        "as": props.value.as,
        "__label": "spinner",
        "__css": spinnerStyles.value
      }, attrs), {
        "default": function _default() {
          return [props.value.label ? vue.createVNode(cVisuallyHidden.CVisuallyHidden, null, {
            "default": function _default() {
              return [props.value.label];
            }
          }) : null];
        }
      });
    };
  }
});
var CSpinner$1 = CSpinner;

exports.CSpinner = CSpinner$1;
