// @chakra-ui/c-accordion/dist/chakra-ui-c-accordion.esm.js
import { defineComponent, ref, computed, createVNode, mergeProps, isVNode } from 'vue';
import * as accordion from '@zag-js/accordion';
import { useMachine, normalizeProps } from '@zag-js/vue';
import { useMultiStyleConfig, StylesProvider, chakra, useStyles } from '@chakra-ui/vue-system';
import { useId } from '@chakra-ui/vue-composables';
import { createContext, SNAO, vueThemingProps, genId, getValidChildren } from '@chakra-ui/vue-utils';
import { mergeWith, filterUndefined } from '@chakra-ui/utils';
import { CCollapse } from '@chakra-ui/c-motion';
import { CIcon } from '@chakra-ui/c-icon';

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

var _excluded = ["hidden"];

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}

var _createContext = createContext({
  name: "AccordionContext",
  strict: true
}),
    AccordionProvider = _createContext[0],
    useAccordion = _createContext[1];

var CAccordion = defineComponent({
  name: "CAccordion",
  props: _extends({
    as: {
      type: [String],
      "default": "div"
    },
    allowMultiple: Boolean,
    allowToggle: Boolean,
    index: SNAO,
    defaultIndex: SNAO,
    reduceMotion: {
      type: Boolean,
      "default": false
    }
  }, vueThemingProps),
  setup: function setup(_props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var uid = ref(genId());
    var context = computed(function () {
      return {
        multiple: _props.allowMultiple,
        collapsible: _props.allowToggle
      };
    });

    var _useMachine = useMachine(accordion.machine({
      id: uid.value
    }), {
      context: context
    }),
        state = _useMachine[0],
        send = _useMachine[1];

    var apiRef = computed(function () {
      return accordion.connect(state.value, send, normalizeProps);
    });
    var props = computed(function () {
      return mergeWith({}, _props, attrs);
    });
    var themingProps = computed(function () {
      return filterUndefined({
        colorScheme: props.value.colorScheme,
        variant: props.value.variant,
        size: props.value.size,
        styleConfig: props.value.styleConfig
      });
    });
    var styles = useMultiStyleConfig("Accordion", themingProps);
    var reduceMotion = computed(function () {
      return props.value.reduceMotion;
    });
    AccordionProvider({
      api: apiRef,
      reduceMotion: reduceMotion
    });
    StylesProvider(styles);
    return function () {
      var api = apiRef.value;
      return createVNode(chakra.div, {
        "sx": {
          "> div": styles.value.root
        }
      }, {
        "default": function _default() {
          return createVNode("div", api.rootProps, [getValidChildren(slots)]);
        }
      });
    };
  }
});

var _createContext2 = createContext({
  name: "AccordionItemContext",
  strict: true
}),
    AccordionItemProvider = _createContext2[0],
    useAccordionItem = _createContext2[1];

var CAccordionItem = defineComponent({
  name: "CAccordionItem",
  props: {
    disabled: Boolean,
    value: String
  },
  setup: function setup(props, _ref2) {
    var slots = _ref2.slots,
        attrs = _ref2.attrs;

    var _uid = useId(undefined, "accordion-item");

    var id = computed(function () {
      return attrs.id || _uid.value;
    });

    var _useAccordion = useAccordion(),
        api = _useAccordion.api;

    var itemValue = computed(function () {
      return id.value;
    });
    var state = computed(function () {
      return api.value.getItemState({
        value: itemValue.value
      });
    });
    var isOpen = computed(function () {
      return state.value.isOpen;
    });
    var isDisabled = computed(function () {
      return props.disabled;
    });
    AccordionItemProvider({
      id: id,
      isOpen: isOpen,
      isDisabled: isDisabled
    });
    var styles = useStyles();
    var containerStyles = computed(function () {
      return _extends({}, styles.value.container, {
        overflowAnchor: "none"
      });
    });
    return function () {
      var _slot;

      return createVNode(chakra.div, mergeProps({
        "__css": containerStyles.value
      }, api.value.getItemProps({
        value: itemValue.value,
        disabled: props.disabled
      }), attrs), _isSlot(_slot = getValidChildren(slots)) ? _slot : {
        "default": function _default() {
          return [_slot];
        }
      });
    };
  }
});
var CAccordionButton = defineComponent({
  name: "CAccordionButton",
  props: {
    disabled: Boolean
  },
  setup: function setup(props, _ref3) {
    var slots = _ref3.slots,
        attrs = _ref3.attrs;

    var _useAccordionItem = useAccordionItem(),
        id = _useAccordionItem.id;

    var _useAccordion2 = useAccordion(),
        api = _useAccordion2.api;

    var styles = useStyles();
    var buttonStyles = computed(function () {
      return _extends({
        display: "flex",
        alignItems: "center",
        width: "100%",
        outline: 0
      }, styles.value.button);
    });
    return function () {
      return createVNode(chakra.button, mergeProps(api.value.getTriggerProps({
        value: id.value,
        disabled: props.disabled
      }), {
        "__css": buttonStyles.value
      }, attrs), {
        "default": function _default() {
          return getValidChildren(slots);
        }
      });
    };
  }
});
var CAccordionPanel = defineComponent({
  name: "CAccordionPanel",
  props: {
    disabled: Boolean
  },
  setup: function setup(props, _ref4) {
    var slots = _ref4.slots,
        attrs = _ref4.attrs;

    var _useAccordionItem2 = useAccordionItem(),
        id = _useAccordionItem2.id,
        isOpen = _useAccordionItem2.isOpen;

    var _useAccordion3 = useAccordion(),
        api = _useAccordion3.api;

    var styles = useStyles();
    return function () {
      var _slot2;

      var _api$value$getContent = api.value.getContentProps({
        value: id.value,
        disabled: props.disabled
      });
          _api$value$getContent.hidden;
          var contentProps = _objectWithoutPropertiesLoose(_api$value$getContent, _excluded);

      return createVNode(CCollapse, {
        "isOpen": isOpen.value
      }, {
        "default": function _default() {
          return [createVNode(chakra.div, mergeProps(contentProps, {
            "__css": styles.value.panel
          }, attrs), _isSlot(_slot2 = getValidChildren(slots)) ? _slot2 : {
            "default": function _default() {
              return [_slot2];
            }
          })];
        }
      });
    };
  }
});
var CAccordionIcon = defineComponent({
  name: "CAccordionIcon",
  setup: function setup(props, _ref5) {
    _ref5.slots;
        var attrs = _ref5.attrs;

    var _useAccordionItem3 = useAccordionItem(),
        isOpen = _useAccordionItem3.isOpen,
        isDisabled = _useAccordionItem3.isDisabled;

    var _useAccordion4 = useAccordion(),
        reduceMotion = _useAccordion4.reduceMotion;

    var styles = useStyles();
    var iconStyles = computed(function () {
      return _extends({
        opacity: isDisabled.value ? 0.4 : 1,
        transform: isOpen.value ? "rotate(-180deg)" : undefined,
        transition: reduceMotion.value ? undefined : "transform 0.2s",
        transformOrigin: "center"
      }, styles.value.icon);
    });
    return function () {
      return createVNode(CIcon, mergeProps({
        "viewBox": "0 0 24 24",
        "aria-hidden": true,
        "__css": iconStyles.value
      }, attrs), {
        "default": function _default() {
          return [createVNode("path", {
            "fill": "currentColor",
            "d": "M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          }, null)];
        }
      });
    };
  }
});

export { CAccordion, CAccordionButton, CAccordionIcon, CAccordionItem, CAccordionPanel };
