// @chakra-ui/c-pin-input/dist/chakra-ui-c-pin-input.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue$1 = require('vue');
var vueSystem = require('@chakra-ui/vue-system');
var vueUtils = require('@chakra-ui/vue-utils');
var cInput = require('@chakra-ui/c-input');
var pinInput = require('@zag-js/pin-input');
var vue = require('@zag-js/vue');

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

var usePinInputMachine = function usePinInputMachine(params, emit) {
  var _useMachine = vue.useMachine(pinInput.machine({
    id: params.id,
    value: params.modelValue || params.value,
    placeholder: params.placeholder,
    type: params.type,
    otp: params.otp,
    mask: params.mask,
    blurOnComplete: params.blurOnComplete,
    onChange: function onChange(value) {
      emit("change", value);
      emit("update:modelValue", value);
    },
    onComplete: function onComplete(_ref) {
      var value = _ref.value,
          valueAsString = _ref.valueAsString;
      emit("complete", {
        value: value,
        valueAsString: valueAsString
      });
    },
    onInvalid: function onInvalid(details) {
      emit("invalid", details);
    }
  })),
      state = _useMachine[0],
      send = _useMachine[1];

  var api = vue$1.computed(function () {
    return pinInput.connect(state.value, send, vue.normalizeProps);
  });
  vue$1.watch(function () {
    return params.modelValue;
  }, function (value) {
    api.value.setValue(Array.from(value));
  });
  return api;
};

var _createStylesContext = vueSystem.createStylesContext("CPinInput"),
    StylesProvider = _createStylesContext[0],
    useStyles = _createStylesContext[1];

var _createContext = vueUtils.createContext(),
    PinInputProvider = _createContext[0],
    usePinInput = _createContext[1];
var CPinInputProps = _extends({
  value: {
    type: Object,
    "default": []
  },
  modelValue: {
    type: Object,
    "default": false
  },
  id: {
    type: String,
    "default": "0"
  },
  placeholder: {
    type: String,
    "default": "o"
  },
  type: {
    type: String,
    "default": "numeric"
  },
  otp: {
    type: Boolean,
    "default": false
  },
  mask: {
    type: Boolean,
    "default": false
  },
  blurOnComplete: {
    type: Boolean,
    "default": false
  },
  dir: {
    type: String,
    "default": "ltr"
  },
  spacing: {
    type: [String, Number],
    "default": "0.75"
  }
}, vueUtils.vueThemingProps);
var CPinInput = vue$1.defineComponent({
  name: "CPinInput",
  props: CPinInputProps,
  emits: ["change", "invalid", "complete", "update:modelValue"],
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs,
        emit = _ref.emit;
    var styles = vueSystem.useStyleConfig("PinInput", props);
    var inputStyles = vue$1.computed(function () {
      return _extends({}, styles.value, {
        mx: props.spacing
      });
    });
    StylesProvider(inputStyles);
    var api = usePinInputMachine(props, emit);
    PinInputProvider(api);
    return function () {
      return vue$1.createVNode(vueSystem.chakra.div, vue$1.mergeProps({
        "__label": "pin-input"
      }, api.value.rootProps, attrs), {
        "default": function _default() {
          return vueUtils.getValidChildren(slots).map(function (child, index) {
            return child.type.name === "CPinInputField" ? vue$1.h(child, {
              index: index
            }) : child;
          });
        }
      });
    };
  }
});
var CPinInputField = vue$1.defineComponent({
  name: "CPinInputField",
  props: {
    index: {
      type: Number,
      "default": 0
    }
  },
  setup: function setup(props, _ref2) {
    var attrs = _ref2.attrs;
    var styles = useStyles();
    var api = usePinInput();
    return function () {
      return vue$1.createVNode(cInput.CInput, vue$1.mergeProps({
        "__label": "pin-input-field",
        "__css": styles.value
      }, api.value.getInputProps({
        index: props.index
      }), attrs), null);
    };
  }
});
var CPinInputClearButton = vue$1.defineComponent({
  name: "CPinInputClearButton",
  setup: function setup(_, _ref3) {
    var slots = _ref3.slots,
        attrs = _ref3.attrs;
    var api = usePinInput();
    return function () {
      return vue$1.createVNode(vueSystem.chakra.button, vue$1.mergeProps({
        "__label": "pin-input-clear-button",
        "onClick": function onClick() {
          return api.value.clearValue();
        }
      }, attrs), {
        "default": function _default() {
          return vueUtils.getValidChildren(slots);
        }
      });
    };
  }
});

exports.CPinInput = CPinInput;
exports.CPinInputClearButton = CPinInputClearButton;
exports.CPinInputField = CPinInputField;
exports.CPinInputProps = CPinInputProps;
exports.PinInputProvider = PinInputProvider;
exports.usePinInput = usePinInput;
